Modified extension gadget

miha-stopar opened the pull request:
### Description

This PR will replace old modified extension node PRs (there were multiple because the witness generator was in a separate repository at that time).

Support for modified extension nodes.

This happens when an extension node is replaced by another (shorter or longer) extension node. For example, we have an extension node at `n1 n2 n3 n4 n5 n6` with branch with two leaves at positions `n` and `m`. If we add a leaf at `n1 n2 n3 n4 n7` where `n5 != n7`, a new extension node is inserted at `n1 n2 n3 n4` with a new branch with an extension node at position `n5` (with only one nibble `n6`) and a leaf at position `n7`. In this case, the S proof contains the extension node at `n1 n2 n3 n4 n5 n6` and no underlying branch and leaf (the modification happens at `n1 n2 n3 n4 n7` and only an extension node is find there), thus we need to add a placeholder branch and a placeholder leaf.

### Type of change

- [x] New feature (non-breaking change which adds functionality)

ChihChengLiang commented on mpt-witness-generator/witness/leaf.go:
```diff
@@ -529,6 +534,11 @@ func prepareStorageLeafNode(leafS, leafC, neighbourNode []byte, storage_key comm
 	}
 	rows = append(rows, nonExistingStorageRow)
 
+	for i := 0; i < 6; i++ {
+		row := make([]byte, valueLen)
+		rows = append(rows, row)
+	}
+
```
Can we explain this?

miha-stopar commented:
> First round of review on the witness generator part.
> I'll review the circuit in the next round.

Thanks @ChihChengLiang ! I think addressed all the comments in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685/commits/67399e9a144929c6f9e90c02bfb7790706bca00a.

miha-stopar commented on zkevm-circuits/src/mpt_circuit/extension.rs:
```diff
@@ -71,7 +71,7 @@ impl<F: Field> ExtensionGadget<F> {
                     meta,
                     cb,
-                    ExtensionBranchRowType::KeyC as usize,
+                    ExtensionBranchRowType::Nibbles as usize,
```
An example extension node witness is:
```
[130 0 150 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
[160 215 121 207 40 14 160 149 115 175 222 139 45 208 88 81 65 226 190 111 191 208 252 147 90 105 163 154 4 132 52 204 121 0]
[0 0 6 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
[160 208 13 115 5 131 97 180 161 37 184 30 176 46 254 113 83 244 235 231 202 172 66 172 99 25 151 249 73 32 160 1 218 0]
```

The third row (`ExtensionBranchRowType::Nibbles`) contains the second nibbles of the key (the first row). So we want to have the range check `0 - 15` which is done if the row is marked by `ExtensionBranchRowType::Nibbles`.

miha-stopar commented on zkevm-circuits/src/mpt_circuit/extension.rs:
```diff
@@ -152,14 +152,6 @@ impl<F: Field> ExtensionGadget<F> {
             // Make sure the nibble counter is updated correctly
             let num_nibbles = key_data.num_nibbles.expr() + num_nibbles.expr();
 
-            // The parity alternates when there's an even number of nibbles, remains the
-            // same otherwise
-            let is_key_odd = ifx! {config.is_key_part_odd => {
-                not!(key_data.is_odd)
-            } elsex {
-                key_data.is_odd.expr()
-            }};
-
```
I moved the parity update down to the line 176 to have the information about the parity _before_ the extension node in `ext_key_rlc_calc_value`. Please see the comment below.

miha-stopar commented on zkevm-circuits/src/mpt_circuit/mod_extension.rs:
```diff
@@ -0,0 +1,327 @@
+
+            let is_insert = parent_data[0].is_placeholder.expr(); // insert or delete
+
+            let lo_s = is_insert.clone() * parent_data[0].hash.lo().expr()
+                + (1.expr() - is_insert.clone()) * parent_data[1].hash.lo().expr();
+            let hi_s = is_insert.clone() * parent_data[0].hash.hi().expr()
+                + (1.expr() - is_insert.clone()) * parent_data[1].hash.hi().expr();
+            let lo_c = is_insert.clone() * parent_data[0].drifted_parent_hash.lo().expr()
+                + (1.expr() - is_insert.clone()) * parent_data[1].drifted_parent_hash.lo().expr();
+            let hi_c = is_insert.clone() * parent_data[0].drifted_parent_hash.hi().expr()
+                + (1.expr() - is_insert.clone()) * parent_data[1].drifted_parent_hash.hi().expr();
```
Yes, totally! Just needed to add `.expr()` after each word, otherwise getting:
```
no method named `mul` found for tuple `(circuit_tools::cell_manager::Cell<F>, circuit_tools::cell_manager::Cell<F>, circuit_tools::cell_manager::Cell<F>, circuit_tools::cell_manager::Cell<F>)`
```

ChihChengLiang commented on zkevm-circuits/src/mpt_circuit/mod_extension.rs:
```diff
@@ -0,0 +1,325 @@
+        FixedTableTag, RlpItemType,
+    },
+};
+
```
Let me know if the doc makes sense
```suggestion
/// Validity conditions for ModExtensionGadget:
/// 1. Both the long and short extension nodes are in the parent node.
/// 2. Both the branches of the short extension node’s and the long extension node’s must be the same
/// 3. The concatenation of the middle extension node and short ext node must equal the long ext node
```

ChihChengLiang reviewed (commented):
This is my final round of review.
I'll approve after the feedback is addressed.

miha-stopar commented on zkevm-circuits/src/mpt_circuit/account_leaf.rs:
```diff
@@ -261,6 +271,17 @@ impl<F: Field> AccountLeafConfig<F> {
                 );
             }
+
+            ifx! {or::expr(&[config.is_mod_extension[0].clone(), config.is_mod_extension[1].clone()]) => {
+                config.mod_extension = ModExtensionGadget::configure(
+                    meta,
+                    cb,
+                    ctx.clone(),
+                    parent_data,
+                    key_data,
+                );
+            }};
```
As expected the block is executed when the condition is true.

For passing the condition to the gadget - that would mean having one big if statement in `ModExtensionGadget` or having the condition in each constraint statement. I think the current code is somehow easier to read?

Brechtpd replied:
Any specific concerns here about the use of `ifx!` @ChihChengLiang ? I think `ifx` is being used here similarly to other places, so not sure how the specific concerns here.

Some of the gadgets are created inside the `ifx!` because otherwise things also get quite a bit messier to keep track of all the different paths and inputs, but regardless of the condition they are always created.

ChihChengLiang replied:
Hi @Brechtpd @miha-stopar,
I've managed to convince myself that the current way of `ifx!` works with this mini example https://github.com/privacy-scaling-explorations/zkevm-circuits/commit/1dd12cadca0094884010724af12dffb8f2887ba7. The concern is resolved. Also thank you all for explaining to me.

Brechtpd commented on zkevm-circuits/src/mpt_circuit/helpers.rs:
```diff
@@ -261,7 +261,7 @@ pub(crate) fn ext_key_rlc_expr<F: Field>(
                 calc_rlc(cb, &key_bytes, 1.expr())
             },
-            and::expr(&[is_long.expr(), is_key_odd.expr()]) => {
+            and::expr(&[is_long.expr(), or::expr(&[and::expr(&[not!(is_key_odd.expr()), not!(is_key_part_odd.expr())]), and::expr(&[is_key_odd.expr(), is_key_part_odd.expr()])])]) => {
```
```suggestion
            and::expr(&[is_long.expr(), not!(xor::expr(is_key_odd.expr(), is_key_part_odd.expr()))]) => {
```

Also helps to make it more clear it's the `else` clause. 

Hmmm maybe it's even better if it's split up in `is_long` and `is_short` cases, and then the `is_long` case contains an `if xor::expr(is_key_odd.expr(), is_key_part_odd.expr()) { } else { }`.

miha-stopar replied:
Changed in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685/commits/4acc57394e92f71d24b4622c78d885a01a698a3d

ChihChengLiang reviewed (approved):
~~The rest parts LGTM. I still have some concerns on [this part](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685#discussion_r1425598669), but plan to create a minimal example to demonstrate the effect after the merging of this PR.~~

Update: LGTM. Concerns resolved in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1685#discussion_r1437140715
